File: security-framework/src/CSRFProtection.ts
import { randomBytes, createHmac } from 'crypto';

export interface CSRFOptions {
  secret?: string;
  tokenName?: string;
  cookieName?: string;
  headerName?: string;
  saltLength?: number;
  skipFailureCallback?: boolean;
  ignoreMethods?: string[];
  value?: (req: any) => string;
}

export interface CSRFToken {
  token: string;
  secret: string;
  timestamp: number;
}

/**
 * CSRF Protection implementation using double-submit cookie pattern
 * Provides secure token generation and validation
 */
export class CSRFProtection {
  private options: Required<CSRFOptions>;
  private readonly defaultSecret: string;

  constructor(options: CSRFOptions = {}) {
    this.defaultSecret = randomBytes(32).toString('hex');
    
    this.options = {
      secret: options.secret || this.defaultSecret,
      tokenName: options.tokenName || '_csrf',
      cookieName: options.cookieName || '_csrf_token',
      headerName: options.headerName || 'X-CSRF-Token',
      saltLength: options.saltLength || 8,
      skipFailureCallback: options.skipFailureCallback ?? false,
      ignoreMethods: options.ignoreMethods || ['GET', 'HEAD', 'OPTIONS'],
      value: options.value || this.defaultTokenExtractor
    };
  }

  /**
   * Generate a new CSRF token
   */
  generateToken(): CSRFToken {
    const salt = randomBytes(this.options.saltLength);
    const secret = this.options.secret;
    const timestamp = Date.now();
    
    // Create token using HMAC with salt and timestamp
    const hmac = createHmac('sha256', secret);
    hmac.update(salt);
    hmac.update(timestamp.toString());
    
    const hash = hmac.digest('hex');
    const token = salt.toString('hex') + hash;
    
    return {
      token,
      secret,
      timestamp
    };
  }

  /**
   * Validate a CSRF token
   */
  validateToken(token: string, maxAge: number = 3600000): boolean { // 1 hour default
    if (!token || typeof token !== 'string') {
      return false;
    }

    try {
      // Extract salt and hash from token
      const saltLength = this.options.saltLength * 2; // hex encoding doubles length
      if (token.length <= saltLength) {
        return false;
      }

      const saltHex = token.substring(0, saltLength);
      const providedHash = token.substring(saltLength);
      
      // Find valid timestamp by trying recent timestamps
      const currentTime = Date.now();
      const startTime = currentTime - maxAge;
      
      // Try timestamps in 1-second intervals
      for (let timestamp = currentTime; timestamp >= startTime; timestamp -= 1000) {
        const expectedHash = this.computeTokenHash(saltHex, timestamp);
        if (this.secureCompare(providedHash, expectedHash)) {
          return true;
        }
      }
      
      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Create CSRF middleware for Express-like frameworks
   */
  middleware() {
    return (req: any, res: any, next: any) => {
      const method = req.method?.toUpperCase();
      
      // Skip validation for safe methods
      if (this.options.ignoreMethods.includes(method)) {
        return next();
      }

      // Generate token if not present
      if (!req.csrfToken) {
        const tokenData = this.generateToken();
        req.csrfToken = () => tokenData.token;
        
        // Set cookie with token
        res.cookie(this.options.cookieName, tokenData.token, {
          httpOnly: true,
          secure: req.secure || req.headers['x-forwarded-proto'] === 'https',
          sameSite: 'strict'
        });
      }

      // For state-changing methods, validate token
      if (!this.options.ignoreMethods.includes(method)) {
        const providedToken = this.extractToken(req);
        const cookieToken = req.cookies?.[this.options.cookieName];
        
        // Double-submit cookie validation
        if (!providedToken || !cookieToken || !this.validateToken(providedToken)) {
          const error = new Error('Invalid CSRF token');
          (error as any).code = 'EBADCSRFTOKEN';
          (error as any).status = 403;
          
          if (this.options.skipFailureCallback) {
            return next();
          }
          return next(error);
        }

        // Additional validation: cookie and form token should match
        if (providedToken !== cookieToken) {
          const error = new Error('CSRF token mismatch');
          (error as any).code = 'EBADCSRFTOKEN';
          (error as any).status = 403;
          return next(error);
        }
      }

      next();
    };
  }

  /**
   * Generate HTML hidden input for forms
   */
  generateFormField(token: string): string {
    return `<input type="hidden" name="${this.options.tokenName}" value="${token}">`;
  }

  /**
   * Generate meta tag for client-side access
   */
  generateMetaTag(token: string): string {
    return `<meta name="csrf-token" content="${token}">`;
  }

  /**
   * Client-side JavaScript code for automatic form enhancement
   */
  generateClientScript(): string {
    return `
      (function() {
        // Get CSRF token from meta tag
        function getCSRFToken() {
          const meta = document.querySelector('meta[name="csrf-token"]');
          return meta ? meta.getAttribute('content') : null;
        }

        // Add CSRF token to all forms
        function enhanceForms() {
          const token = getCSRFToken();
          if (!token) return;

          const forms = document.querySelectorAll('form');
          forms.forEach(form => {
            // Skip if already has CSRF field
            if (form.querySelector('input[name="${this.options.tokenName}"]')) return;
            
            // Skip GET forms
            if (form.method.toLowerCase() === 'get') return;

            // Add hidden CSRF field
            const input = document.createElement('input');
            input.type = 'hidden';
            input.name = '${this.options.tokenName}';
            input.value = token;
            form.appendChild(input);
          });
        }

        // Enhance fetch requests
        const originalFetch = window.fetch;
        window.fetch = function(url, options = {}) {
          const token = getCSRFToken();
          if (token && options.method && !['GET', 'HEAD', 'OPTIONS'].includes(options.method.toUpperCase())) {
            options.headers = options.headers || {};
            options.headers['${this.options.headerName}'] = token;
          }
          return originalFetch.call(this, url, options);
        };

        // Enhance XMLHttpRequest
        const originalOpen = XMLHttpRequest.prototype.open;
        const originalSend = XMLHttpRequest.prototype.send;
        
        XMLHttpRequest.prototype.open = function(method, url, async, user, password) {
          this._method = method;
          return originalOpen.call(this, method, url, async, user, password);
        };
        
        XMLHttpRequest.prototype.send = function(data) {
          const token = getCSRFToken();
          if (token && this._method && !['GET', 'HEAD', 'OPTIONS'].includes(this._method.toUpperCase())) {
            this.setRequestHeader('${this.options.headerName}', token);
          }
          return originalSend.call(this, data);
        };

        // Run enhancement on DOM ready
        if (document.readyState === 'loading') {
          document.addEventListener('DOMContentLoaded', enhanceForms);
        } else {
          enhanceForms();
        }

        // Re-enhance when new content is added
        const observer = new MutationObserver(enhanceForms);
        observer.observe(document.body, { childList: true, subtree: true });
      })();
    `;
  }

  private extractToken(req: any): string | null {
    // Try to extract token from various sources
    return this.options.value(req) ||
           req.body?.[this.options.tokenName] ||
           req.query?.[this.options.tokenName] ||
           req.headers?.[this.options.headerName.toLowerCase()] ||
           null;
  }

  private defaultTokenExtractor(req: any): string | null {
    return req.body?._csrf || 
           req.query?._csrf ||
           req.headers?.['x-csrf-token'] ||
           null;
  }

  private computeTokenHash(saltHex: string, timestamp: number): string {
    const hmac = createHmac('sha256', this.options.secret);
    hmac.update(Buffer.from(saltHex, 'hex'));
    hmac.update(timestamp.toString());
    return hmac.digest('hex');
  }

  private secureCompare(a: string, b: string): boolean {
    if (a.length !== b.length) {
      return false;
    }

    let result = 0;
    for (let i = 0; i < a.length; i++) {
      result |= a.charCodeAt(i) ^ b.charCodeAt(i);
    }

    return result === 0;
  }

  /**
   * Rotate the secret key (for key rotation security)
   */
  rotateSecret(): string {
    const newSecret = randomBytes(32).toString('hex');
    this.options.secret = newSecret;
    return newSecret;
  }

  /**
   * Get current configuration
   */
  getConfig(): Readonly<Required<CSRFOptions>> {
    return { ...this.options };
  }

  /**
   * Create a validator function for custom implementations
   */
  createValidator(maxAge?: number) {
    return (token: string) => this.validateToken(token, maxAge);
  }
}

File: compiler/src/csrf_protection.rs
use anyhow::Result;
use rand::Rng;
use sha2::{Sha256, Digest};
use std::time::{SystemTime, UNIX_EPOCH};

/// CSRF Protection Implementation
/// Uses double-submit cookie strategy for stateless CSRF protection
pub struct CsrfProtection {
    secret_key: String,
    token_length: usize,
    cookie_name: String,
    header_name: String,
}

impl CsrfProtection {
    /// Create new CSRF protection instance
    pub fn new(secret_key: String) -> Self {
        Self {
            secret_key,
            token_length: 32,
            cookie_name: "_eghact_csrf".to_string(),
            header_name: "X-CSRF-Token".to_string(),
        }
    }

    /// Generate a cryptographically secure CSRF token
    pub fn generate_token(&self) -> Result<CsrfToken> {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)?
            .as_secs();
        
        // Generate random bytes
        let mut rng = rand::thread_rng();
        let random_bytes: Vec<u8> = (0..self.token_length)
            .map(|_| rng.gen())
            .collect();
        
        // Create token data
        let token_data = format!("{}:{}", timestamp, hex::encode(&random_bytes));
        
        // Generate HMAC signature
        let signature = self.generate_hmac(&token_data)?;
        let full_token = format!("{}:{}", token_data, signature);
        
        // Encode as base64 for safe transmission
        let encoded_token = base64::encode(full_token);
        
        Ok(CsrfToken {
            value: encoded_token,
            timestamp,
            expires_in: 3600, // 1 hour default
        })
    }

    /// Validate CSRF token from double-submit cookie
    pub fn validate_token(&self, cookie_token: &str, submitted_token: &str) -> Result<bool> {
        // Tokens must match (double-submit check)
        if cookie_token != submitted_token {
            return Ok(false);
        }

        // Decode and validate token structure
        let decoded = base64::decode(submitted_token)
            .map_err(|_| anyhow::anyhow!("Invalid token encoding"))?;
        
        let token_str = String::from_utf8(decoded)
            .map_err(|_| anyhow::anyhow!("Invalid token format"))?;
        
        let parts: Vec<&str> = token_str.split(':').collect();
        if parts.len() != 3 {
            return Ok(false);
        }

        let timestamp: u64 = parts[0].parse()
            .map_err(|_| anyhow::anyhow!("Invalid timestamp"))?;
        let random_part = parts[1];
        let signature = parts[2];

        // Check token expiration
        let current_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)?
            .as_secs();
        
        if current_time - timestamp > 3600 { // 1 hour expiration
            return Ok(false);
        }

        // Verify HMAC signature
        let token_data = format!("{}:{}", timestamp, random_part);
        let expected_signature = self.generate_hmac(&token_data)?;
        
        // Constant-time comparison to prevent timing attacks
        Ok(constant_time_eq(signature.as_bytes(), expected_signature.as_bytes()))
    }

    /// Generate JavaScript for client-side CSRF handling
    pub fn generate_client_code(&self) -> String {
        format!(r#"
// Eghact CSRF Protection Client Library
(function() {{
    const CSRF_COOKIE_NAME = '{}';
    const CSRF_HEADER_NAME = '{}';
    
    // Get CSRF token from cookie
    function getCsrfToken() {{
        const cookies = document.cookie.split(';');
        for (let cookie of cookies) {{
            const [name, value] = cookie.trim().split('=');
            if (name === CSRF_COOKIE_NAME) {{
                return decodeURIComponent(value);
            }}
        }}
        return null;
    }}
    
    // Set CSRF token in form
    function setCsrfTokenInForm(form) {{
        const token = getCsrfToken();
        if (!token) {{
            console.warn('CSRF token not found');
            return false;
        }}
        
        // Remove existing CSRF input
        const existingInput = form.querySelector('input[name="_csrf"]');
        if (existingInput) {{
            existingInput.remove();
        }}
        
        // Add CSRF token as hidden input
        const input = document.createElement('input');
        input.type = 'hidden';
        input.name = '_csrf';
        input.value = token;
        form.appendChild(input);
        
        return true;
    }}
    
    // Intercept form submissions
    function protectForms() {{
        document.addEventListener('submit', function(event) {{
            const form = event.target;
            if (form.method && form.method.toLowerCase() === 'post') {{
                if (!setCsrfTokenInForm(form)) {{
                    event.preventDefault();
                    console.error('CSRF protection failed: token not available');
                    return false;
                }}
            }}
        }});
    }}
    
    // Protect fetch requests
    const originalFetch = window.fetch;
    window.fetch = function(url, options = {{}}) {{
        const token = getCsrfToken();
        
        // Add CSRF header for POST, PUT, DELETE, PATCH requests
        if (options.method && ['POST', 'PUT', 'DELETE', 'PATCH'].includes(options.method.toUpperCase())) {{
            if (token) {{
                options.headers = options.headers || {{}};
                options.headers[CSRF_HEADER_NAME] = token;
            }} else {{
                console.warn('CSRF token not found for fetch request');
            }}
        }}
        
        return originalFetch.call(this, url, options);
    }};
    
    // Initialize CSRF protection
    if (document.readyState === 'loading') {{
        document.addEventListener('DOMContentLoaded', protectForms);
    }} else {{
        protectForms();
    }}
    
    // Export functions for manual use
    window.EghactCSRF = {{
        getToken: getCsrfToken,
        protectForm: setCsrfTokenInForm,
        headerName: CSRF_HEADER_NAME,
        cookieName: CSRF_COOKIE_NAME
    }};
}})();
"#, self.cookie_name, self.header_name)
    }

    /// Generate middleware for development server
    pub fn generate_middleware_code(&self) -> String {
        format!(r#"
// Eghact CSRF Protection Middleware
const crypto = require('crypto');
const {{ parse: parseCookies }} = require('cookie');

const CSRF_COOKIE_NAME = '{}';
const CSRF_HEADER_NAME = '{}';
const SECRET_KEY = process.env.EGHACT_CSRF_SECRET || 'dev-secret-change-in-production';

function generateToken() {{
    const timestamp = Math.floor(Date.now() / 1000);
    const randomBytes = crypto.randomBytes(16).toString('hex');
    const tokenData = `${{timestamp}}:${{randomBytes}}`;
    
    const hmac = crypto.createHmac('sha256', SECRET_KEY);
    hmac.update(tokenData);
    const signature = hmac.digest('hex');
    
    const fullToken = `${{tokenData}}:${{signature}}`;
    return Buffer.from(fullToken).toString('base64');
}}

function validateToken(cookieToken, submittedToken) {{
    // Double-submit check
    if (cookieToken !== submittedToken) {{
        return false;
    }}
    
    try {{
        const decoded = Buffer.from(submittedToken, 'base64').toString('utf8');
        const parts = decoded.split(':');
        
        if (parts.length !== 3) {{
            return false;
        }}
        
        const [timestamp, randomPart, signature] = parts;
        const currentTime = Math.floor(Date.now() / 1000);
        
        // Check expiration (1 hour)
        if (currentTime - parseInt(timestamp) > 3600) {{
            return false;
        }}
        
        // Verify signature
        const tokenData = `${{timestamp}}:${{randomPart}}`;
        const hmac = crypto.createHmac('sha256', SECRET_KEY);
        hmac.update(tokenData);
        const expectedSignature = hmac.digest('hex');
        
        return crypto.timingSafeEqual(
            Buffer.from(signature, 'hex'),
            Buffer.from(expectedSignature, 'hex')
        );
    }} catch (error) {{
        return false;
    }}
}}

function csrfMiddleware(req, res, next) {{
    // Generate and set CSRF token for all requests
    const cookies = parseCookies(req.headers.cookie || '');
    let csrfToken = cookies[CSRF_COOKIE_NAME];
    
    if (!csrfToken) {{
        csrfToken = generateToken();
        res.setHeader('Set-Cookie', 
            `${{CSRF_COOKIE_NAME}}=${{csrfToken}}; HttpOnly; SameSite=Strict; Path=/`
        );
    }}
    
    // Validate CSRF for state-changing methods
    if (['POST', 'PUT', 'DELETE', 'PATCH'].includes(req.method)) {{
        const submittedToken = req.headers[CSRF_HEADER_NAME.toLowerCase()] || req.body._csrf;
        
        if (!validateToken(csrfToken, submittedToken)) {{
            res.status(403).json({{
                error: 'CSRF token validation failed',
                code: 'CSRF_INVALID'
            }});
            return;
        }}
    }}
    
    // Make token available to templates
    req.csrfToken = csrfToken;
    next();
}}

module.exports = {{
    csrfMiddleware,
    generateToken,
    validateToken,
    CSRF_COOKIE_NAME,
    CSRF_HEADER_NAME
}};
"#, self.cookie_name, self.header_name)
    }

    /// Generate HMAC signature for token
    fn generate_hmac(&self, data: &str) -> Result<String> {
        let mut hasher = Sha256::new();
        hasher.update(format!("{}:{}", self.secret_key, data));
        Ok(hex::encode(hasher.finalize()))
    }
}

/// CSRF Token structure
#[derive(Debug, Clone)]
pub struct CsrfToken {
    pub value: String,
    pub timestamp: u64,
    pub expires_in: u64,
}

impl CsrfToken {
    /// Check if token is expired
    pub fn is_expired(&self) -> Result<bool> {
        let current_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)?
            .as_secs();
        
        Ok(current_time - self.timestamp > self.expires_in)
    }
    
    /// Get token value for use in forms/headers
    pub fn value(&self) -> &str {
        &self.value
    }
}

/// Constant-time string comparison to prevent timing attacks
fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    
    let mut result = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        result |= x ^ y;
    }
    
    result == 0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_token_generation() {
        let csrf = CsrfProtection::new("test-secret".to_string());
        let token1 = csrf.generate_token().unwrap();
        let token2 = csrf.generate_token().unwrap();
        
        // Tokens should be different
        assert_ne!(token1.value, token2.value);
        
        // Tokens should be valid base64
        assert!(base64::decode(&token1.value).is_ok());
        assert!(base64::decode(&token2.value).is_ok());
    }

    #[test]
    fn test_token_validation() {
        let csrf = CsrfProtection::new("test-secret".to_string());
        let token = csrf.generate_token().unwrap();
        
        // Valid double-submit should pass
        assert!(csrf.validate_token(&token.value, &token.value).unwrap());
        
        // Different tokens should fail
        let token2 = csrf.generate_token().unwrap();
        assert!(!csrf.validate_token(&token.value, &token2.value).unwrap());
        
        // Malformed token should fail
        assert!(!csrf.validate_token(&token.value, "invalid").unwrap());
    }

    #[test]
    fn test_constant_time_eq() {
        assert!(constant_time_eq(b"hello", b"hello"));
        assert!(!constant_time_eq(b"hello", b"world"));
        assert!(!constant_time_eq(b"hello", b"hell"));
        assert!(!constant_time_eq(b"hell", b"hello"));
    }
}